c_src/erlang_jq_port.c

```diff
@@ -364,25 +392,83 @@ static bool handle_set_filter_program_lru_cache_size() {
 static bool handle_get_filter_program_lru_cache_size() {
     char buffer[64];
     sprintf(buffer, "%d", erlang_jq_get_filter_program_lru_cache_size());
-    if (write_packet((byte*)buffer, strlen(buffer)) <= 0) {
+    if (write_packet(STDOUT_FILENO, (byte*)buffer, strlen(buffer)) <= 0) {
         return false;
     }
     return true;
 }
 
 static bool handle_ping() {
     const char* response = "pong";
-    if (write_packet((byte*)response, strlen(response)) <= 0) {
+    if (write_packet(STDOUT_FILENO, (byte*)response, strlen(response)) <= 0) {
         return false;
     }
     return true;
 }
 
+static void child_dead_sig_handler(int signum){
+    LOG_PRINT("Child is dead\n");
+    // Normal exit is not considered safe in signal handlers
+    _Exit(0);
+}
+
+static void input_forwarder_porcess() {
+    // We are in the child process
+    set_erljq_alloc(jq_port_alloc);
+    set_erljq_free(free);
+    byte* packet;
+    uint64_t size;
+    while (true) {
+        packet = read_whole_packet(STDIN_FILENO, &size);
+        if (packet == NULL) {
+            LOG_PRINT("STDIN closed. Erlang VM is dead or port is closed. Worker process will get SIGCHLD.\n");
+            break;
+        }
+        ssize_t write_exact_res = write_exact(pipe_out, packet, size);
```

Comment:
this will never run if L422 is blocking? consider the following scenario:

- start vm and vm starts port program
- port program fork this `forwarder`
- kill main worker
- this forwarder will stay?

Reply:
Yes, it will stay until port program is closed on the Erlang end or a message is sent to the port. If a message is sent to the port program, the forwarder will die with due to SIGPIPE when writing to the pipe... Hmm, I think you have a good point. If the the main process is killed brutally (e.g., SIGKILL) then the pipe might not be closed which is not good. I think this can be solved by trapping signals that can kill the processes and make sure that the pipe is closed in the signal handlers, right?

Reply:
fixed in commit below by using threads instead of forking the process.